### PR TITLE
feat: add global state hook

### PR DESCRIPTION
**Pull request by MrHuangJser**
# Description

## Type of change

- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [x] New feature _(non-breaking change which adds functionality)_
- [ ] **Breaking change** _(fix or feature that would cause existing functionality to not work as before)_

# Checklist
- [x] Read the [Contributing Guide](https://github.com/streamich/react-use/blob/master/CONTRIBUTING.md)
- [x] Perform a code self-review
- [x] Comment the code, particularly in hard-to-understand areas
- [x] Add documentation
- [x] Add hook's story at Storybook
- [x] Cover changes with tests
- [x] Ensure the test suite passes (`yarn test`)
- [x] Provide 100% tests coverage
- [x] Make sure code lints (`yarn lint`). Fix it with `yarn lint:fix` in case of failure.
- [x] Make sure types are fine (`yarn lint:types`).

### CODE REVIEW

File: src/useGlobalState.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      store.state = state;
+      store.setters.forEach(setter => setter(store.state));
+    },
+    setters: [],
+  };
+
+  return (): [S | undefined, (state: S) => void] => {
+    const [globalState, stateSetter] = useState<S | undefined>(store.state);
+
+    if (!store.setters.includes(stateSetter)) {
```
Comment (streamich) [inline src/useGlobalState.ts]: Seems this `IfStatement` will be called on every re-render. Probably better to put it inside `useLayoutEffect`?

Comment (MrHuangJser) [reply src/useGlobalState.ts]: yeah, i'll pull a new request

File: docs/createGlobalState.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# `useGlobalState`
+
+A React hook witch create a globally shared state.
```
Comment (streamich) [inline docs/createGlobalState.md]:
```suggestion
A React hook which creates a globally shared state.
```

**Comment by mbest** [comment]
This PR moved the following code

```ts
      if (!store.setters.includes(stateSetter)) {
        store.setters.push(stateSetter);
      }
```

to be within a `useLayoutEffect`. Are you sure that `useLayoutEffect` is synchronous? If not, you could have a race condition where the value is changed after you read the value but before you subscribe to changes.

**Comment by MrHuangJser** [comment]
> In this PR, you moved the following code
> 
> ```ts
>       if (!store.setters.includes(stateSetter)) {
>         store.setters.push(stateSetter);
>       }
> ```
> 
> to be within a `useLayoutEffect`. Are you sure that `useLayoutEffect` is synchronous? If not, you could have a race condition where the value is changed after you read the value but before you subscribe to changes.

thk  your replay i'll fix improve it in next pr :)

**Pull request merged**